lttoolbox/transducer.cc

```diff
@@ -1395,3 +1395,204 @@ Transducer::applyACX(Alphabet& alpha,
+        // label becomes input of f and output of g
+        if (g_left != 0 && // we've already dealt with g epsilons
+            f_a.sameSymbol(f_output, g_a, g_left, true)) {
+          next = std::make_pair(f_trg, g_trg);
+          if (seen.find(next) == seen.end()) {
+            todo.push_back(next);
+          }
+          if (states_f_g_gf.find(next) == states_f_g_gf.end()) {
+            states_f_g_gf.insert(std::make_pair(next, gf.newState()));
+          }
+          int gf_trg = states_f_g_gf[next];
+          int32_t gf_label;
+          if (g_right >= 0) { // non-symbols are equal across alphabets
+            gf_label = f_inverted ? f_a(g_right, f_input) : f_a(f_input, g_right);
+          } else {
+            UString g_rightstr;
+            g_a.getSymbol(g_rightstr, g_right);
+            gf_label = f_inverted ? f_a(f_a(g_rightstr), f_input) : f_a(f_input, f_a(g_rightstr));
+          }
+          gf.linkStates(gf_src,   // fromState
+                        gf_trg,   // toState
+                        gf_label, // symbol-pair, using f alphabet
+                        f_wt + g_wt); // weight of transduction – composition adds weights!
+        }
+      }
+      if(g_anywhere && g_src == g.initial) {
+        // If g_anywhere, all g entries are optional – we always add
+        // the transitions that were already in f:
+        next = std::make_pair(f_trg, g_src);
+        if (seen.find(next) == seen.end()) {
+          todo.push_back(next);
+        }
+        if (states_f_g_gf.find(next) == states_f_g_gf.end()) {
+          states_f_g_gf.insert(std::make_pair(next, gf.newState()));
+        }
+        int gf_trg = states_f_g_gf[next];
+        gf.linkStates(gf_src,  // fromState
+                      gf_trg,  // toState
+                      f_label, // symbol-pair, using f alphabet
+                      f_wt);   // weight of transduction
+      }
+      // If f has an epsilon, also add a transition not to g.initial but g_src:
+      if(f_output == 0) {      // will be the left if f_inverted
+        next = std::make_pair(f_trg, g_src);
+        if (seen.find(next) == seen.end()) {
+          todo.push_back(next);
+        }
+        if (states_f_g_gf.find(next) == states_f_g_gf.end()) {
+          states_f_g_gf.insert(std::make_pair(next, gf.newState()));
+        }
+        int gf_trg = states_f_g_gf[next];
+        gf.linkStates(gf_src,  // fromState
+                      gf_trg,  // toState
+                      f_label, // symbol-pair, using f alphabet
+                      f_wt);   // weight of transduction
+      }
+    } // end loop arcs from f_src
+  } // end while todo
+
+  for(auto& it : states_f_g_gf)
+  {
+    int s_f = it.first.first;
+    int s_g = it.first.second;
+    int s_gf = it.second;
+    if(isFinal(s_f) && (g.isFinal(s_g)
+                        || (g_anywhere && g.initial == s_g)))
+    {
+      gf.finals.insert(std::make_pair(s_gf, default_weight));
```

Comment:
this will discard any final weights from the input fsts. I think the weight should be `f.finals[s_f] + g.finals[s_g]`

also it took me a bit to figure out why `(g_anywhere && g.initial == s_g)` is the condition we want for final `?*`, so maybe add (if this matches your understanding) `// if we're in anywhere mode, every state will be paired with g.initial if it's not paired with something in the middle of g`

Reply:
copied from https://github.com/unhammer/lttoolbox/blob/compose/lttoolbox/transducer.cc#L1274  – is this perhaps buggy too? Somehow I didn't consider weights on final states as a possibility :)

Reply:
It wouldn't surprise me. There's several existing bugs with our handling of weights (see #44, for the biggest one).